### PR TITLE
Gen4 Planner: AxB vs BxA

### PR DESCRIPTION
**Pull request by systay**
This PR teaches the Gen4 planner how to re-evaluate routes after a predicate has been pushed during the join tree comparisons.

### Better plans

I'll use an example to show what the planner does:

```sql
SELECT user.col 
FROM user JOIN user_extra ON user.id = user_extra.col
```

There is a primary vindex on `user.id`, but for `user_extra`, there are no applicable vindexes.
This means first of all that we can't merge the two tables into a single route - for certain `user.id` values, the matching `user_extra.col` might well be placed in a different shard.

So, we are forced to do the join on the vtgate - we can't delegate this to mysql. Now the question is: should we first query the `user` table and push down the `user.id` to the `user_extra` route, or the other way around?

Given the available vindexes, we have to evaluate `user_extra` using a scatter query no matter if we have `user.id` available or not. For the `user` table on the other hand, if we have `user_extra.col` available, we could use a `SelectEqualUnique` which is way better than a scatter query.

### Implementation comments

Because the planner now takes the same predicate (`user.id = user_extra.col`) and rewrites it in two different ways depending on which route is on the LHS, I needed to make it possible to clone expressions. I have hand-written the clone method for expressions, but I think it makes sense to use the visitorgen to generate this code, and do it for all AST structs and not just `Expr`.

I also had to move the pushing of predicates to the joinTree phase instead of waiting until we have logical plans. 

### Planner performance

I was afraid that all this new logic that is being done during joinTree selection would lead to much slower planning, but the benchmarks don't look too scary. 😅 

```
BenchmarkPlanner/from_cases.txt-v3-16       	  	    	    1198	    853119 ns/op	  393968 B/op	    7319 allocs/op
BenchmarkPlanner/from_cases.txt-v4-16         	    		    1436	    796068 ns/op	  366321 B/op	    7553 allocs/op
BenchmarkPlanner/from_cases.txt-v4left2right-16         	    1545	    747779 ns/op	  342363 B/op	    6999 allocs/op
BenchmarkPlanner/filter_cases.txt-v3-16                 	    1544	    763470 ns/op	  340849 B/op	    6820 allocs/op
BenchmarkPlanner/filter_cases.txt-v4-16                 	    1749	    656536 ns/op	  291623 B/op	    5779 allocs/op
BenchmarkPlanner/filter_cases.txt-v4left2right-16       	    1764	    658813 ns/op	  288869 B/op	    5705 allocs/op
BenchmarkPlanner/large_cases.txt-v3-16                  	    8616	    126044 ns/op	   54688 B/op	    1189 allocs/op
BenchmarkPlanner/large_cases.txt-v4-16                  	    6630	    163335 ns/op	   72650 B/op	    1564 allocs/op
BenchmarkPlanner/large_cases.txt-v4left2right-16        	   13299	     88614 ns/op	   41088 B/op	    1003 allocs/op
BenchmarkPlanner/aggr_cases.txt-v3-16                   	    2864	    433544 ns/op	  148570 B/op	    4068 allocs/op
BenchmarkPlanner/aggr_cases.txt-v4-16                   	    3622	    337122 ns/op	  110093 B/op	    3157 allocs/op
BenchmarkPlanner/aggr_cases.txt-v4left2right-16         	    3675	    336505 ns/op	  110174 B/op	    3157 allocs/op
BenchmarkPlanner/select_cases.txt-v3-16                 	    1092	    967935 ns/op	  328761 B/op	    8525 allocs/op
BenchmarkPlanner/select_cases.txt-v4-16                 	    1335	    911256 ns/op	  285991 B/op	    7859 allocs/op
BenchmarkPlanner/select_cases.txt-v4left2right-16       	    1269	    900011 ns/op	  279665 B/op	    7769 allocs/op
BenchmarkPlanner/union_cases.txt-v3-16                  	    1315	    948248 ns/op	  338814 B/op	    9025 allocs/op
BenchmarkPlanner/union_cases.txt-v4-16                  	    1221	    925400 ns/op	  338622 B/op	    9025 allocs/op
BenchmarkPlanner/union_cases.txt-v4left2right-16        	    1280	    899773 ns/op	  338773 B/op	    9025 allocs/op
BenchmarkPlanner/wireup_cases.txt-v3-16                 	   19597	     58601 ns/op	   24786 B/op	     586 allocs/op
BenchmarkPlanner/wireup_cases.txt-v4-16                 	   25574	     45954 ns/op	   18512 B/op	     470 allocs/op
BenchmarkPlanner/wireup_cases.txt-v4left2right-16       	   27835	     42890 ns/op	   15732 B/op	     429 allocs/op
```

### CODE REVIEW

**Review by harshit-gangal — COMMENTED** [review]
code looks good to be. I think there is need to document the logic for the important methods involved in the planning.

File: go/vt/vtgate/planbuilder/testdata/filter_cases.txt [hunk 1]
```diff
@@ -981,7 +996,44 @@ Gen4 plan same as above
...
+          "Sharded": true
+        },
+        "FieldQuery": "select 1 from user where 1 != 1",
+        "Query": "select 1 from user where user.id = :unsharded_id",
+        "Table": "user",
+        "Values": [
+          ":unsharded_id"
+        ],
+        "Vindex": "user_index"
+      }
```
[the remark this answers is not in the document]
Comment (systay) [reply go/vt/vtgate/planbuilder/testdata/filter_cases.txt:1033]: maybe. would you mind if we look deeper into this later, when a larger percentage of the gen4 planner is done, and we can get a better overview of the options available?

File: go/vt/vtgate/planbuilder/route_planning_test.go [hunk 2]
```diff
@@ -104,3 +104,23 @@ func TestMergeJoins(t *testing.T) {
 		})
 	}
 }
+
```
[the remark this answers is not in the document]
Comment (systay) [reply go/vt/vtgate/planbuilder/route_planning_test.go:107]: should probably add to it. it was not a conscious decision, no

File: go/vt/vtgate/planbuilder/route_planning.go [hunk 3]
```diff
@@ -147,54 +153,94 @@ type (
...
 
 // cost implements the joinTree interface
-func (*routePlan) cost() int {
+func (rp *routePlan) cost() int {
+	switch rp.routeOpCode {
+	case
+		engine.SelectDBA,
+		engine.SelectEqualUnique,
+		engine.SelectNext,
+		engine.SelectNone,
```
[the remark this answers is not in the document]
Comment (systay) [reply go/vt/vtgate/planbuilder/route_planning.go:209]: 👍

File: go/vt/vtgate/planbuilder/route_planning.go [hunk 4]
```diff
@@ -203,30 +249,64 @@ func (rp *routePlan) addPredicate(predicates ...sqlparser.Expr) error {
...
+	rp.predicates = append(rp.predicates, predicates...)
+
+	return nil
+}
+
+// pickBestAvailableVindex goes over the available vindexes for this route and picks the best one available.
+func (rp *routePlan) pickBestAvailableVindex() {
 	//TODO (Manan,Andres): Improve cost metric for vindexes
-	for _, v := range vindexPreds {
+	for _, v := range rp.vindexPreds {
```
[the remark this answers is not in the document]
Comment (systay) [reply go/vt/vtgate/planbuilder/route_planning.go]: the idea is that when we add a table to a routePlan, we also add all possible vindexes. later, when predicates are added, new vindexes can become available (in other words, `covered` as the struct field).

I tried removing vindexes that had already lost and did not need rechecking, but that didn't make any diff in the benchmarks, so I didn't keep that complexity in here

**Review by harshit-gangal — APPROVED** [review]

**Review by harshit-gangal — APPROVED** [review]

**Pull request merged**